reference/test/reorder/mc64_kernels.cpp

```diff
@@ -0,0 +1,401 @@
+    GKO_ASSERT_NEAR(row_scaling[5], value_type{2. / 15.}, this->tolerance);
+    GKO_ASSERT_NEAR(col_scaling[0], value_type{1.}, this->tolerance);
+    GKO_ASSERT_NEAR(col_scaling[1], value_type{1.5}, this->tolerance);
+    GKO_ASSERT_NEAR(col_scaling[2], value_type{0.9375}, this->tolerance);
+    GKO_ASSERT_NEAR(col_scaling[3], value_type{5. / 6.}, this->tolerance);
+    GKO_ASSERT_NEAR(col_scaling[4], value_type{1.}, this->tolerance);
+    GKO_ASSERT_NEAR(col_scaling[5], value_type{1.25}, this->tolerance);
+}
+
+
+TYPED_TEST(Mc64, CreatesCorrectPermutationAndScalingLargeTrivialExampleProduct)
+{
+    using index_type = typename TestFixture::index_type;
+    using real_type = typename TestFixture::real_type;
+    using value_type = typename TestFixture::value_type;
+    using matrix_type = typename TestFixture::matrix_type;
+
+    std::ifstream mtx_stream{gko::matrices::location_1138_bus_mtx};
+    auto mtx = gko::share(gko::read<matrix_type>(mtx_stream, this->ref));
+    std::ifstream result_stream{gko::matrices::location_1138_bus_mc64_result};
+    auto expected_result = gko::read<matrix_type>(result_stream, this->ref);
+
+    auto mc64_factory =
+        gko::reorder::Mc64<value_type, index_type>::build()
+            .with_strategy(gko::reorder::mc64_strategy::max_diagonal_product)
+            .on(this->ref);
+    auto mc64 = mc64_factory->generate(mtx);
+
+    auto perm = mc64->get_permutation();
+    auto row_scaling = mc64->get_row_scaling();
+    auto col_scaling = mc64->get_col_scaling();
+
+    col_scaling->rapply(mtx.get(), mtx.get());
+    row_scaling->apply(mtx.get(), mtx.get());
+    perm->apply(mtx.get(), mtx.get());
+
+    GKO_ASSERT_MTX_NEAR(mtx, expected_result, this->tolerance);
+}
+
+
+TYPED_TEST(Mc64, CreatesCorrectPermutationAndScalingLargeExampleProduct)
+{
+    using index_type = typename TestFixture::index_type;
+    using real_type = typename TestFixture::real_type;
+    using value_type = typename TestFixture::value_type;
+    using matrix_type = typename TestFixture::matrix_type;
+
+    std::ifstream mtx_stream{gko::matrices::location_nontrivial_mc64_example};
+    auto mtx = gko::share(gko::read<matrix_type>(mtx_stream, this->ref));
+    mtx->sort_by_column_index();
+    std::ifstream result_stream{gko::matrices::location_nontrivial_mc64_result};
+    auto expected_result = gko::read<matrix_type>(result_stream, this->ref);
+
+    auto mc64_factory =
+        gko::reorder::Mc64<value_type, index_type>::build()
+            .with_strategy(gko::reorder::mc64_strategy::max_diagonal_product)
+            .on(this->ref);
+    auto mc64 = mc64_factory->generate(mtx);
+
+    auto perm = mc64->get_permutation();
+    auto row_scaling = mc64->get_row_scaling();
+    auto col_scaling = mc64->get_col_scaling();
+
+    col_scaling->rapply(mtx.get(), mtx.get());
+    row_scaling->apply(mtx.get(), mtx.get());
+    perm->apply(mtx.get(), mtx.get());
+
+    GKO_ASSERT_MTX_NEAR(mtx, expected_result, 1e-6);
```

Comment:
why to use 1e-6?

Reply:
The tolerance is because the result is compared against output from Matlab's `equilibrate`, which does compute the same permutation and scaling coefficients, but uses `log` instead of `log2` in the algorithm.